File: barretenberg/test/wire_compat/cpp/echo_server.cpp
/**
 * Echo IPC server (C++) — uses GENERATED types + template IPC server.
 * Usage: echo_server --socket /tmp/echo.sock
 */

#include "generated/types_gen.hpp"
#include "generated/ipc_server.hpp"
#include <iostream>
#include <string_view>

using namespace echo;

int main(int argc, char** argv) {
    const char* socket_path = nullptr;
    for (int i = 1; i < argc - 1; i++) {
        if (std::string_view(argv[i]) == "--socket") {
            socket_path = argv[i + 1];
        }
    }
    if (!socket_path) {
        std::cerr << "Usage: echo_server --socket <path>\n";
        return 1;
    }

    // Serve using template IPC server + generated types for dispatch
    ipc::serve(socket_path, [](const std::vector<uint8_t>& payload) -> std::vector<uint8_t> {
        auto oh = msgpack::unpack(reinterpret_cast<const char*>(payload.data()), payload.size());
        auto obj = oh.get();
        auto& inner = obj.via.array.ptr[0];
        std::string cmd_name(inner.via.array.ptr[0].via.str.ptr, inner.via.array.ptr[0].via.str.size);
        auto& cmd_payload = inner.via.array.ptr[1];

        msgpack::sbuffer resp_buf;
        msgpack::packer<msgpack::sbuffer> pk(resp_buf);

        // Echo dispatch using GENERATED types
        if (cmd_name == "EchoBytes") {
            EchoBytes cmd; cmd_payload.convert(cmd);
            pk.pack_array(2); pk.pack(std::string("EchoBytesResponse")); pk.pack(cmd);
        } else if (cmd_name == "EchoFields") {
            EchoFields cmd; cmd_payload.convert(cmd);
            pk.pack_array(2); pk.pack(std::string("EchoFieldsResponse")); pk.pack(cmd);
        } else if (cmd_name == "EchoNested") {
            EchoNested cmd; cmd_payload.convert(cmd);
            pk.pack_array(2); pk.pack(std::string("EchoNestedResponse")); pk.pack(cmd);
        } else if (cmd_name == "EchoShutdown") {
            pk.pack_array(2); pk.pack(std::string("EchoShutdownResponse")); pk.pack_map(0);
        } else {
            pk.pack_array(2); pk.pack(std::string("EchoErrorResponse"));
            pk.pack_map(1); pk.pack(std::string("message")); pk.pack(std::string("Unknown: ") + cmd_name);
        }

        return std::vector<uint8_t>(resp_buf.data(), resp_buf.data() + resp_buf.size());
    });

    return 0;
}

File: barretenberg/test/wire_compat/zig/echo_server.zig
/// Echo IPC server (Zig) — uses GENERATED types + IPC server template.
/// Hand-written dispatch + echo handlers.
/// Usage: echo_server --socket /tmp/echo.sock
const std = @import("std");
const msgpack = @import("msgpack");
const Payload = msgpack.Payload;
const types = @import("generated/echo_types.zig");
const ipc_server = @import("generated/ipc_server.zig");

const alloc = std.heap.page_allocator;

pub fn main() !void {
    var args = std.process.args();
    _ = args.next();
    var socket_path: ?[]const u8 = null;
    while (args.next()) |arg| {
        if (std.mem.eql(u8, arg, "--socket")) {
            socket_path = args.next();
        }
    }
    const path = socket_path orelse {
        std.debug.print("Usage: echo_server --socket <path>\n", .{});
        return error.InvalidArgument;
    };

    try ipc_server.serve(path, dispatch);
}

fn dispatch(cmd_name: []const u8, fields: Payload) ipc_server.DispatchResult {
    // Shutdown
    if (std.mem.eql(u8, cmd_name, "EchoShutdown")) {
        return .{ .resp_name = "EchoShutdownResponse", .resp_payload = Payload.mapPayload(alloc) };
    }

    // EchoBytes — echo back
    if (std.mem.eql(u8, cmd_name, "EchoBytes")) {
        const cmd = types.EchoBytes.fromPayload(fields) catch return makeError("deser failed");
        const resp = types.EchoBytesResponse{ .data = cmd.data };
        return .{ .resp_name = "EchoBytesResponse", .resp_payload = resp.toPayload(alloc) };
    }

    // EchoFields — echo back
    if (std.mem.eql(u8, cmd_name, "EchoFields")) {
        const cmd = types.EchoFields.fromPayload(fields) catch return makeError("deser failed");
        const resp = types.EchoFieldsResponse{ .a = cmd.a, .b = cmd.b, .name = cmd.name };
        return .{ .resp_name = "EchoFieldsResponse", .resp_payload = resp.toPayload(alloc) };
    }

    // EchoNested — echo back
    if (std.mem.eql(u8, cmd_name, "EchoNested")) {
        const cmd = types.EchoNested.fromPayload(fields) catch return makeError("deser failed");
        const resp = types.EchoNestedResponse{ .inner = cmd.inner };
        return .{ .resp_name = "EchoNestedResponse", .resp_payload = resp.toPayload(alloc) };
    }

    return makeError("unknown command");
}

fn makeError(message: []const u8) ipc_server.DispatchResult {
    var err_map = Payload.mapPayload(alloc);
    err_map.mapPut("message", Payload.strToPayload(message, alloc) catch return .{ .resp_name = "EchoErrorResponse", .resp_payload = Payload.mapPayload(alloc) }) catch {};
    return .{ .resp_name = "EchoErrorResponse", .resp_payload = err_map };
}
